cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <deque>
+#include <span>
+
+#include <rapidsmpf/streaming/core/channel.hpp>
+#include <rapidsmpf/streaming/core/coro_utils.hpp>
+#include <rapidsmpf/streaming/core/fanout.hpp>
+#include <rapidsmpf/streaming/core/message.hpp>
+#include <rapidsmpf/streaming/core/node.hpp>
+
+#include <coro/coro.hpp>
+
+namespace rapidsmpf::streaming::node {
+namespace {
+
+/**
+ * @brief Try to allocate memory from the memory types that the message content uses.
+ *
+ * @param msg The message to allocate memory for.
+ * @return The memory types to try to allocate from.
+ */
+constexpr std::span<const MemoryType> try_memory_types(Message const& msg) {
+    auto const& cd = msg.content_description();
+    // if the message content uses device memory, try to allocate from device memory
+    // first, else allocate from host memory
+    return cd.content_size(MemoryType::DEVICE) > 0
+               ? MEMORY_TYPES
+               : std::span<const MemoryType>{
+                     MEMORY_TYPES.begin() + static_cast<std::size_t>(MemoryType::HOST),
+                     MEMORY_TYPES.end()
+                 };
+}
+
+/**
+ * @brief Asynchronously send a message to multiple output channels.
+ *
+ * @param msg The message to broadcast. Each channel receives a deep copy of the original
+ * message.
+ * @param chs_out The set of output channels to which the message is sent.
+ */
+Node send_to_channels(
+    Context* ctx, Message&& msg, std::vector<std::shared_ptr<Channel>>& chs_out
+) {
+    RAPIDSMPF_EXPECTS(!chs_out.empty(), "output channels cannot be empty");
+
+    std::vector<coro::task<bool>> tasks;
+    tasks.reserve(chs_out.size());
+    for (size_t i = 0; i < chs_out.size() - 1; i++) {
+        // do a reservation for each copy, so that it will fallback to host memory if
+        // needed
+        auto res = ctx->br()->reserve_or_fail(msg.copy_cost(), try_memory_types(msg));
+        tasks.emplace_back(chs_out[i]->send(msg.copy(res)));
+    }
```

Comment:
*question*: Should this copy and send pattern be scheduled? Otherwise we serialise all the copying and spilling.